### PR TITLE
fix forwarding with clinet authentication  over TLS (forward to gmail)

### PR DESCRIPTION
**Pull request by ayankel**
No new features.

Fixes Incorrect sequence of events (AUTH and STARTTLS) when operate as forwarding proxy when upstream server requires both authentication and secure connection.  also fix extra \r\n that choke gmail .

Running tests:  [ 'tests/plugins/queue/smtp_forward.js' ]
OK: 11 assertions (253ms)

### CODE REVIEW

File: plugins/queue/smtp_forward.js [hunk 1]
```diff
@@ -70,7 +70,15 @@ exports.hook_queue = function (next, connection) {
 
         if (cfg.auth_user) {
             connection.loginfo(plugin, 'Configuring authentication for SMTP server ' + cfg.host + ':' + cfg.port);
-            smtp_client.on('capabilities', function () {
+            smtp_client.on('capabilities', function() {
+                connection.loginfo(plugin, 'capabilities received');
+                if ('secured' in smtp_client) {
+                    connection.loginfo(plugin, 'secured is pending');
+                    if (smtp_client.secured === false) {
+                        connection.loginfo(plugin,"Waiting for STARTTLS to complete. AUTH posponed");
```
Comment (msimerson) [inline plugins/queue/smtp_forward.js]: typo: posponed -> postponed

File: smtp_client.js [hunk 2]
```diff
@@ -360,7 +369,7 @@ exports.get_client_plugin = function (plugin, connection, c, callback) {
     pool.acquire(function (err, smtp_client) {
         connection.logdebug(plugin, 'Got smtp_client: ' + smtp_client.uuid);
 
-        var secured = false;
+        var secured = false;
```
Comment (msimerson) [inline smtp_client.js]: you've added some trailing whitespace that should be removed.

File: smtp_client.js [hunk 3]
```diff
@@ -382,7 +391,7 @@ exports.get_client_plugin = function (plugin, connection, c, callback) {
 
         var helo = function (command) {
             if (smtp_client.xclient) {
-                smtp_client.send_command(command, connection.hello.host);
+                smtp_client.send_command(command, connection.hello_host);
             }
```
Comment (msimerson) [inline smtp_client.js]: hello_host is deprecated, please revert this to use `connection.hello.host`

File: smtp_client.js [hunk 4]
```diff
@@ -394,12 +403,13 @@ exports.get_client_plugin = function (plugin, connection, c, callback) {
...
                 secured = true;
+                smtp_client.secured = true;
                 smtp_client.emit('greeting', 'EHLO');
             };
             for (var line in smtp_client.response) {
                 if (smtp_client.response[line].match(/^XCLIENT/)) {
                     if (!smtp_client.xclient) {
-                        smtp_client.send_command('XCLIENT', 'ADDR=' + connection.remote.ip);
+                        smtp_client.send_command('XCLIENT', 'ADDR=' + connection.remote_ip);
                         return;
```
Comment (msimerson) [inline smtp_client.js]: same as above. use `connection.remote.ip`

File: smtp_client.js [hunk 5]
```diff
@@ -484,7 +499,7 @@ exports.get_client_plugin = function (plugin, connection, c, callback) {
 
         if (smtp_client.connected) {
             if (smtp_client.xclient) {
-                smtp_client.send_command('XCLIENT', 'ADDR=' + connection.remote.ip);
+                smtp_client.send_command('XCLIENT', 'ADDR=' + connection.remote_ip);
             }
```
Comment (msimerson) [inline smtp_client.js]: same as above.

File: smtp_client.js [hunk 6]
```diff
@@ -65,14 +66,22 @@ function SMTPClient (port, host, connect_timeout, idle_timeout) {
             return;
         }
 
-        if (client.command === 'auth') {
-            if (code.match(/^3/) && cont === 'VXNlcm5hbWU6') {
+        if (client.command === 'auth' || client.authenticating) {
+            logger.loginfo('SERVER RESPONCE, CLIENT ' + client.command + ", authenticating=" + client.authenticating + ",code="+code + ",cont="+cont+",msg=" +msg);
```
Comment (msimerson) [inline smtp_client.js]: spelling error: RESPONCE -> RESPONSE

File: plugins/queue/smtp_forward.js [hunk 7]
```diff
@@ -82,15 +90,19 @@ exports.hook_queue = function (next, connection) {
...
                 }
                 else if (cfg.auth_type === 'login') {
+                    smtp_client.authenticating = true;
+                    smtp_client.authenticated=false;
+
+                    connection.loginfo(plugin, 'Authenticating with AUTH LOGIN ' + cfg.auth_user);
                     smtp_client.send_command('AUTH', 'LOGIN');
-                    smtp_client.on('auth', function () {
-                        connection.loginfo(plugin, 'Authenticating with AUTH LOGIN ' + cfg.auth_user);
+                    smtp_client.on('auth', function() {
```
Comment (msimerson) [inline plugins/queue/smtp_forward.js]: code style issue, please add space between function keyword and parens in function declarations.

**Comment by ayankel** [comment]
updated as requested

File: smtp_client.js [hunk 8]
```diff
@@ -360,7 +369,7 @@ exports.get_client_plugin = function (plugin, connection, c, callback) {
     pool.acquire(function (err, smtp_client) {
         connection.logdebug(plugin, 'Got smtp_client: ' + smtp_client.uuid);
 
-        var secured = false;
+        var secured = false;
```
Comment (msimerson) [inline smtp_client.js]: last issue, you've still got some trailing whitespace here.

Comment (msimerson) [inline smtp_client.js]: That's why the tests are failing:

````
/home/travis/build/haraka/Haraka/smtp_client.js
  372:29   error  Trailing spaces not allowed  no-trailing-spaces
  423:156  error  Trailing spaces not allowed  no-trailing-spaces
````

**Comment by msimerson** [comment]
Looks good @ayankel , thanks!

**Pull request merged**